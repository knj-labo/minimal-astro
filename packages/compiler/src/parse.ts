import type {
  FragmentNode,
  Node,
  ElementNode,
  ComponentNode,
  TextNode,
  ExpressionNode,
  FrontmatterNode,
  Attr,
  Diagnostic,
  ParseResult,
  SourceSpan,
} from '../types/ast.js';
import { Tokenizer, TokenType, type Token } from './tokenizer.js';

export interface ParseOptions {
  filename?: string;
}

export class Parser {
  private tokens: Token[] = [];
  private current: number = 0;
  private diagnostics: Diagnostic[] = [];
  private filename: string;

  constructor(tokens: Token[], options: ParseOptions = {}) {
    this.tokens = tokens;
    this.filename = options.filename || '<anonymous>';
  }

  private peek(offset: number = 0): Token | null {
    const index = this.current + offset;
    return index < this.tokens.length ? this.tokens[index] : null;
  }

  private advance(): Token | null {
    if (this.current < this.tokens.length) {
      return this.tokens[this.current++];
    }
    return null;
  }

  private isAtEnd(): boolean {
    const token = this.peek();
    return !token || token.type === TokenType.EOF;
  }

  private addDiagnostic(
    code: string,
    message: string,
    loc: SourceSpan,
    severity: 'error' | 'warning' = 'error'
  ): void {
    this.diagnostics.push({
      code,
      message,
      loc,
      severity,
    });
  }

  private isComponentTag(name: string): boolean {
    return /^[A-Z]/.test(name);
  }

  private isVoidElement(tag: string): boolean {
    const voidElements = [
      'area', 'base', 'br', 'col', 'embed', 'hr', 'img', 'input',
      'link', 'meta', 'param', 'source', 'track', 'wbr'
    ];
    return voidElements.includes(tag.toLowerCase());
  }

  private parseAttributes(): Attr[] {
    const attrs: Attr[] = [];
    const seenNames = new Set<string>();

    while (!this.isAtEnd()) {
      const token = this.peek();
      if (!token) break;

      if (token.type === TokenType.TagClose || token.type === TokenType.TagSelfClose) {
        break;
      }

      if (token.type === TokenType.AttributeName) {
        const attrToken = this.advance()!;
        const name = attrToken.value;
        let value: string | boolean = true;

        // Check for duplicate client directives
        if (name.startsWith('client:') && seenNames.has(name)) {
          this.addDiagnostic(
            'duplicate-directive',
            `Duplicate ${name} directive`,
            attrToken.loc,
            'warning'
          );
        }
        seenNames.add(name);

        // Check if next token is an attribute value
        const nextToken = this.peek();
        if (nextToken && nextToken.type === TokenType.AttributeValue) {
          const valueToken = this.advance()!;
          const fullValue = valueToken.value;
          const equalIndex = fullValue.indexOf('=');
          if (equalIndex !== -1) {
            value = fullValue.substring(equalIndex + 1);
            // Remove quotes if present
            if ((value.startsWith('"') && value.endsWith('"')) ||
                (value.startsWith("'") && value.endsWith("'"))) {
              value = value.slice(1, -1);
            }
          }
        }

        attrs.push({
          name,
          value,
          loc: attrToken.loc,
        });
      } else {
        // Skip unknown tokens in tag context
        this.advance();
      }
    }

    return attrs;
  }

  private parseFrontmatter(): FrontmatterNode | null {
    const token = this.peek();
    if (token && token.type === TokenType.FrontmatterContent) {
      this.advance();
      return {
        type: 'Frontmatter',
        code: token.value,
        loc: token.loc,
      };
    }
    return null;
  }

  private parseExpression(): ExpressionNode | null {
    const token = this.peek();
    if (token && token.type === TokenType.ExpressionContent) {
      this.advance();
      
      // Check if expression was incomplete (unclosed brace)
      const incomplete = !token.value.includes('}') && this.isAtEnd();
      if (incomplete) {
        this.addDiagnostic(
          'unclosed-expression',
          'Unclosed expression',
          token.loc
        );
      }

      return {
        type: 'Expression',
        code: token.value,
        loc: token.loc,
        ...(incomplete && { incomplete: true }),
      };
    }
    return null;
  }

  private parseText(): TextNode | null {
    const token = this.peek();
    if (token && token.type === TokenType.Text) {
      this.advance();
      return {
        type: 'Text',
        value: token.value,
        loc: token.loc,
      };
    }
    return null;
  }

  private parseElement(): ElementNode | ComponentNode | null {
    const openToken = this.peek();
    if (!openToken || openToken.type !== TokenType.TagOpen) {
      return null;
    }

    this.advance();
    const tag = openToken.value;
    const isComponent = this.isComponentTag(tag);
    const attrs = this.parseAttributes();
    
    let selfClosing = false;
    let children: Node[] = [];

    const closeToken = this.peek();
    if (closeToken) {
      if (closeToken.type === TokenType.TagSelfClose) {
        this.advance();
        selfClosing = true;
      } else if (closeToken.type === TokenType.TagClose) {
        this.advance();
        
        // Parse children if not self-closing and not void element
        if (!this.isVoidElement(tag)) {
          children = this.parseChildren(tag);
        }
      }
    }

    const endLoc = this.tokens[this.current - 1]?.loc.end || openToken.loc.end;
    const loc: SourceSpan = {
      start: openToken.loc.start,
      end: endLoc,
    };

    if (isComponent) {
      return {
        type: 'Component',
        tag,
        attrs,
        children,
        selfClosing,
        loc,
      };
    } else {
      return {
        type: 'Element',
        tag,
        attrs,
        children,
        selfClosing,
        loc,
      };
    }
  }

  private parseChildren(parentTag: string): Node[] {
    const children: Node[] = [];

    while (!this.isAtEnd()) {
      const token = this.peek();
      if (!token) break;

      // Check for closing tag
      if (token.type === TokenType.TagClose && token.value === parentTag) {
        this.advance();
        break;
      }

      // Check for mismatched closing tag
      if (token.type === TokenType.TagClose) {
        this.addDiagnostic(
          'mismatched-tag',
          `Expected closing tag for <${parentTag}> but found </${token.value}>`,
          token.loc
        );
        // Try to recover by treating it as if parent was closed
        break;
      }

      const node = this.parseNode();
      if (node) {
        children.push(node);
      } else {
        // Skip unknown tokens
        this.advance();
      }
    }

    return children;
  }

  private parseNode(): Node | null {
    const expression = this.parseExpression();
    if (expression) return expression;

    const element = this.parseElement();
    if (element) return element;

    const text = this.parseText();
    if (text) return text;

    return null;
  }

  public parse(): ParseResult {
    const children: Node[] = [];
    
    // Check for frontmatter first
    const frontmatter = this.parseFrontmatter();
    if (frontmatter) {
      children.push(frontmatter);
    }

    // Parse remaining content
    while (!this.isAtEnd()) {
      const node = this.parseNode();
      if (node) {
        children.push(node);
      } else {
        // Skip unknown tokens
        this.advance();
      }
    }

    const startLoc = children[0]?.loc.start || { line: 1, column: 1, offset: 0 };
    const endLoc = children[children.length - 1]?.loc.end || startLoc;

    const ast: FragmentNode = {
      type: 'Fragment',
      children,
      loc: {
        start: startLoc,
        end: endLoc,
      },
    };

    return {
      ast,
      diagnostics: this.diagnostics,
    };
  }
}

export function parseAstro(
  source: string,
  options?: ParseOptions
): ParseResult {
  const tokenizer = new Tokenizer(source);
  const tokens = tokenizer.tokenize();
  const parser = new Parser(tokens, options);
  return parser.parse();
}