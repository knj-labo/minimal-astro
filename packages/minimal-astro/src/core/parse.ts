import type {
	Attr,
	ComponentNode,
	Diagnostic,
	ElementNode,
	ExpressionNode,
	FragmentNode,
	FrontmatterNode,
	Node,
	ParseResult,
	SourceSpan,
	TextNode,
} from "../types/ast.js";
import { type Token, TokenType, tokenize } from "./tokenizer.js";
import { safeExecute } from "./utils/error-boundary.js";

export interface ParseOptions {
	filename?: string;
}

interface ParserState {
	readonly tokens: readonly Token[];
	readonly current: number;
	readonly diagnostics: readonly Diagnostic[];
	readonly filename: string;
}

function createInitialState(
	tokens: Token[],
	options: ParseOptions = {},
): ParserState {
	return {
		tokens,
		current: 0,
		diagnostics: [],
		filename: options.filename ?? "<anonymous>",
	};
}

function peek(state: ParserState, offset = 0): Token | null {
	const index = state.current + offset;
	return index < state.tokens.length ? state.tokens[index] : null;
}

function advance(state: ParserState): [ParserState, Token | null] {
	if (state.current < state.tokens.length) {
		const token = state.tokens[state.current];
		return [{ ...state, current: state.current + 1 }, token];
	}
	return [state, null];
}

function isAtEnd(state: ParserState): boolean {
	const token = peek(state);
	return !token || token.type === TokenType.EOF;
}

function addDiagnostic(
	state: ParserState,
	code: string,
	message: string,
	loc: SourceSpan,
	severity: "error" | "warning" = "error",
): ParserState {
	return {
		...state,
		diagnostics: [
			...state.diagnostics,
			{
				code,
				message,
				loc,
				severity,
			},
		],
	};
}

function isComponentTag(name: string): boolean {
	return /^[A-Z]/.test(name);
}

function isVoidElement(tag: string): boolean {
	const voidElements = [
		"area",
		"base",
		"br",
		"col",
		"embed",
		"hr",
		"img",
		"input",
		"link",
		"meta",
		"param",
		"source",
		"track",
		"wbr",
	];
	return voidElements.includes(tag.toLowerCase());
}

function parseAttributes(state: ParserState): [ParserState, Attr[]] {
	const attrs: Attr[] = [];
	const seenNames = new Set<string>();
	let currentState = state;

	while (!isAtEnd(currentState)) {
		const token = peek(currentState);
		if (!token) break;

		if (
			token.type === TokenType.TagClose ||
			token.type === TokenType.TagSelfClose
		) {
			break;
		}

		if (token.type === TokenType.AttributeName) {
			const [newState, attrToken] = advance(currentState);
			currentState = newState;

			if (!attrToken) continue;

			const name = attrToken.value;
			let value: string | boolean = true;

			// Check for duplicate client directives
			if (name.startsWith("client:") && seenNames.has(name)) {
				currentState = addDiagnostic(
					currentState,
					"duplicate-directive",
					`Duplicate ${name} directive`,
					attrToken.loc,
					"warning",
				);
			}
			seenNames.add(name);

			// Check if next token is an attribute value that belongs to this attribute
			const nextToken = peek(currentState);
			if (nextToken && nextToken.type === TokenType.AttributeValue) {
				// If the AttributeValue token contains an '=' and doesn't start with our attribute name,
				// it's a separate attribute, not the value of this one
				const fullValue = nextToken.value;
				const equalIndex = fullValue.indexOf("=");

				if (equalIndex !== -1) {
					const tokenAttrName = fullValue.substring(0, equalIndex);
					// Only consume this token if it belongs to our current attribute
					if (tokenAttrName === name) {
						const [valueState, valueToken] = advance(currentState);
						currentState = valueState;

						if (valueToken) {
							value = fullValue.substring(equalIndex + 1);
							// Remove quotes if present
							if (
								(value.startsWith('"') && value.endsWith('"')) ||
								(value.startsWith("'") && value.endsWith("'"))
							) {
								value = value.slice(1, -1);
							}
						}
					}
					// If tokenAttrName !== name, don't consume the token - let it be processed separately
				}
			}

			attrs.push({
				name,
				value,
				loc: attrToken.loc,
			});
		} else if (token.type === TokenType.AttributeValue) {
			// Handle combined attribute tokens from tokenizer
			const [newState, attrToken] = advance(currentState);
			currentState = newState;

			if (!attrToken) continue;

			const fullValue = attrToken.value;
			const equalIndex = fullValue.indexOf("=");

			let name: string;
			let value: string | boolean = true;

			if (equalIndex !== -1) {
				name = fullValue.substring(0, equalIndex);
				value = fullValue.substring(equalIndex + 1);

				// Remove quotes if present
				if (
					(value.startsWith('"') && value.endsWith('"')) ||
					(value.startsWith("'") && value.endsWith("'"))
				) {
					value = value.slice(1, -1);
				}
			} else {
				name = fullValue;
			}

			// Check for duplicate client directives
			if (name.startsWith("client:") && seenNames.has(name)) {
				currentState = addDiagnostic(
					currentState,
					"duplicate-directive",
					`Duplicate ${name} directive`,
					attrToken.loc,
					"warning",
				);
			}
			seenNames.add(name);

			attrs.push({
				name,
				value,
				loc: attrToken.loc,
			});
		} else {
			// Skip unknown tokens in tag context
			[currentState] = advance(currentState);
		}
	}

	return [currentState, attrs];
}

function parseFrontmatter(
	state: ParserState,
): [ParserState, FrontmatterNode | null] {
	const token = peek(state);
	if (token && token.type === TokenType.FrontmatterContent) {
		const [newState] = advance(state);
		return [
			newState,
			{
				type: "Frontmatter",
				code: token.value,
				loc: token.loc,
			},
		];
	}
	return [state, null];
}

function parseExpression(
	state: ParserState,
): [ParserState, ExpressionNode | null] {
	const token = peek(state);
	if (token && token.type === TokenType.ExpressionContent) {
		const [newState] = advance(state);

		// Check if expression was marked as incomplete
		let code = token.value;
		let incomplete = false;
		let currentState = newState;

		if (code.includes("\x00incomplete")) {
			code = code.replace("\x00incomplete", "");
			incomplete = true;
			currentState = addDiagnostic(
				currentState,
				"unclosed-expression",
				"Unclosed expression",
				token.loc,
			);
		}

		return [
			currentState,
			{
				type: "Expression",
				code,
				loc: token.loc,
				...(incomplete && { incomplete: true }),
			},
		];
	}
	return [state, null];
}

function parseText(state: ParserState): [ParserState, TextNode | null] {
	const token = peek(state);
	if (token && token.type === TokenType.Text) {
		const [newState] = advance(state);
		return [
			newState,
			{
				type: "Text",
				value: token.value,
				loc: token.loc,
			},
		];
	}
	return [state, null];
}

function getImplicitlyClosedTags(tag: string): string[] {
	// HTML tags that are implicitly closed by certain other tags
	const implicitClosers: Record<string, string[]> = {
		li: ["li"],
		dt: ["dt", "dd"],
		dd: ["dt", "dd"],
		p: [
			"p",
			"div",
			"h1",
			"h2",
			"h3",
			"h4",
			"h5",
			"h6",
			"ul",
			"ol",
			"pre",
			"blockquote",
		],
		option: ["option", "optgroup"],
		optgroup: ["optgroup"],
		thead: ["tbody", "tfoot"],
		tbody: ["tbody", "tfoot"],
		tfoot: ["tbody"],
		tr: ["tr"],
		td: ["td", "th"],
		th: ["td", "th"],
	};
	return implicitClosers[tag.toLowerCase()] ?? [];
}

function parseChildren(
	state: ParserState,
	parentTag: string,
): [ParserState, Node[]] {
	const children: Node[] = [];
	const implicitlyClosedBy = getImplicitlyClosedTags(parentTag);
	let currentState = state;

	while (!isAtEnd(currentState)) {
		const token = peek(currentState);
		if (!token) break;

		// Check for closing tag
		if (token.type === TokenType.TagClose && token.value === parentTag) {
			const [newState] = advance(currentState);
			return [newState, children];
		}

		// Check for mismatched closing tag
		if (token.type === TokenType.TagClose) {
			currentState = addDiagnostic(
				currentState,
				"mismatched-tag",
				`Expected closing tag for <${parentTag}> but found </${token.value}>`,
				token.loc,
			);
			// Try to recover by treating it as if parent was closed
			return [currentState, children];
		}

		// Check for tags that implicitly close the parent (like <li> closing previous <li>)
		if (
			token.type === TokenType.TagOpen &&
			implicitlyClosedBy.includes(token.value)
		) {
			// Don't consume the token, let parent handle it
			return [currentState, children];
		}

		const [nodeState, node] = parseNode(currentState);
		currentState = nodeState;
		if (node) {
			children.push(node);
		} else {
			// Skip unknown tokens
			[currentState] = advance(currentState);
		}
	}

	// Check if we reached end without closing tag
	if (isAtEnd(currentState) && !isVoidElement(parentTag)) {
		const lastToken = currentState.tokens[currentState.tokens.length - 1];
		currentState = addDiagnostic(
			currentState,
			"unclosed-tag",
			`Unclosed tag <${parentTag}>`,
			lastToken?.loc ?? {
				start: { line: 1, column: 1, offset: 0 },
				end: { line: 1, column: 1, offset: 0 },
			},
		);
	}

	return [currentState, children];
}

function parseElement(
	state: ParserState,
): [ParserState, ElementNode | ComponentNode | null] {
	const openToken = peek(state);
	if (!openToken || openToken.type !== TokenType.TagOpen) {
		return [state, null];
	}

	let [currentState] = advance(state);
	const tag = openToken.value;
	const isComponent = isComponentTag(tag);
	const [attrState, attrs] = parseAttributes(currentState);
	currentState = attrState;

	let selfClosing = false;
	let children: Node[] = [];

	const closeToken = peek(currentState);
	if (closeToken) {
		if (closeToken.type === TokenType.TagSelfClose) {
			[currentState] = advance(currentState);
			selfClosing = true;
		} else if (closeToken.type === TokenType.TagClose) {
			[currentState] = advance(currentState);

			// Parse children if not self-closing and not void element
			if (!isVoidElement(tag)) {
				const [childrenState, parsedChildren] = parseChildren(
					currentState,
					tag,
				);
				currentState = childrenState;
				children = parsedChildren;
			}
		}
	}

	const endLoc =
		currentState.tokens[currentState.current - 1]?.loc.end ?? openToken.loc.end;
	const loc: SourceSpan = {
		start: openToken.loc.start,
		end: endLoc,
	};

	const node = isComponent
		? {
				type: "Component" as const,
				tag,
				attrs,
				children,
				selfClosing,
				loc,
			}
		: {
				type: "Element" as const,
				tag,
				attrs,
				children,
				selfClosing,
				loc,
			};

	return [currentState, node];
}

function parseNode(state: ParserState): [ParserState, Node | null] {
	// Try parsing in order of likelihood
	const [exprState, expression] = parseExpression(state);
	if (expression) return [exprState, expression];

	const [elemState, element] = parseElement(state);
	if (element) return [elemState, element];

	const [textState, text] = parseText(state);
	if (text) return [textState, text];

	return [state, null];
}

function parse(state: ParserState): ParseResult {
	const children: Node[] = [];
	let currentState = state;

	// Check for frontmatter first
	const [frontmatterState, frontmatter] = parseFrontmatter(currentState);
	if (frontmatter) {
		children.push(frontmatter);
		currentState = frontmatterState;
	}

	// Parse remaining content
	while (!isAtEnd(currentState)) {
		const [nodeState, node] = parseNode(currentState);
		if (node) {
			children.push(node);
			currentState = nodeState;
		} else {
			// Skip unknown tokens
			[currentState] = advance(currentState);
		}
	}

	const startLoc = children[0]?.loc.start ?? { line: 1, column: 1, offset: 0 };
	const endLoc = children[children.length - 1]?.loc.end ?? startLoc;

	const ast: FragmentNode = {
		type: "Fragment",
		children,
		loc: {
			start: startLoc,
			end: endLoc,
		},
	};

	return {
		ast,
		diagnostics: currentState.diagnostics as Diagnostic[],
	};
}

export function parseAstro(
	source: string,
	options?: ParseOptions,
): ParseResult {
	return safeExecute(
		() => {
			const tokens = tokenize(source);
			const initialState = createInitialState(tokens, options);
			return parse(initialState);
		},
		{
			operation: "parse",
			filename: options?.filename,
			context: { sourceLength: source.length },
		},
		{
			fallbackValue: {
				ast: {
					type: "Fragment",
					children: [],
					loc: {
						start: { line: 1, column: 1, offset: 0 },
						end: { line: 1, column: 1, offset: 0 },
					},
				},
				diagnostics: [],
			},
		},
	);
}
