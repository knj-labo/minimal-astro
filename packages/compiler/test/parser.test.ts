import { parseAstro } from '../src/parse.js';
import { readFileSync } from 'fs';
import { join, dirname } from 'path';
import { fileURLToPath } from 'url';
import type { Node, FragmentNode, ElementNode, ComponentNode, TextNode, ExpressionNode, FrontmatterNode } from '../types/ast.js';

const __dirname = dirname(fileURLToPath(import.meta.url));

function loadFixture(name: string): string {
  return readFileSync(join(__dirname, 'fixtures', name), 'utf-8');
}

describe('parseAstro', () => {
  describe('basic parsing', () => {
    test('should parse basic astro file with frontmatter and components', () => {
      const source = loadFixture('basic.astro');
      const result = parseAstro(source);
      
      expect(result.diagnostics).toHaveLength(0);
      expect(result.ast.type).toBe('Fragment');
      expect(result.ast.children).toHaveLength(2);
      
      // Check frontmatter
      const frontmatter = result.ast.children[0] as FrontmatterNode;
      expect(frontmatter.type).toBe('Frontmatter');
      expect(frontmatter.code).toContain("import Layout");
      expect(frontmatter.code).toContain("const title = 'Hello'");
      
      // Check Layout component
      const layout = result.ast.children[1] as ComponentNode;
      expect(layout.type).toBe('Component');
      expect(layout.tag).toBe('Layout');
      expect(layout.children).toHaveLength(7); // Includes whitespace text nodes
      
      // Find h1 element
      const h1 = layout.children.find(child => 
        child.type === 'Element' && (child as ElementNode).tag === 'h1'
      ) as ElementNode;
      expect(h1).toBeDefined();
      expect(h1.children).toHaveLength(1);
      expect(h1.children[0].type).toBe('Expression');
      expect((h1.children[0] as ExpressionNode).code).toBe('title');
      
      // Find p element with client:visible
      const p = layout.children.find(child => 
        child.type === 'Element' && (child as ElementNode).tag === 'p'
      ) as ElementNode;
      expect(p).toBeDefined();
      expect(p.attrs).toHaveLength(1);
      expect(p.attrs[0].name).toBe('client:visible');
      expect(p.attrs[0].value).toBe(true);
      
      // Find Counter component
      const counter = layout.children.find(child => 
        child.type === 'Component' && (child as ComponentNode).tag === 'Counter'
      ) as ComponentNode;
      expect(counter).toBeDefined();
      expect(counter.attrs).toHaveLength(2);
      expect(counter.attrs[0].name).toBe('client:load');
      expect(counter.attrs[1].name).toBe('count');
      expect(counter.attrs[1].value).toBe('{5}');
    });
  });

  describe('self-closing tags', () => {
    test('should handle HTML void elements and component self-closing tags', () => {
      const source = loadFixture('self-closing.astro');
      const result = parseAstro(source);
      
      expect(result.diagnostics).toHaveLength(0);
      
      const div = result.ast.children[0] as ElementNode;
      expect(div.type).toBe('Element');
      expect(div.tag).toBe('div');
      
      // Find img (void element)
      const img = div.children.find(child => 
        child.type === 'Element' && (child as ElementNode).tag === 'img'
      ) as ElementNode;
      expect(img).toBeDefined();
      expect(img.selfClosing).toBe(true);
      expect(img.attrs[0].name).toBe('src');
      expect(img.attrs[0].value).toBe('/logo.svg');
      
      // Find Chart component
      const chart = div.children.find(child => 
        child.type === 'Component' && (child as ComponentNode).tag === 'Chart'
      ) as ComponentNode;
      expect(chart).toBeDefined();
      expect(chart.selfClosing).toBe(true);
      expect(chart.attrs[0].name).toBe('client:idle');
    });
  });

  describe('error recovery', () => {
    test('should recover from unclosed expressions', () => {
      const source = loadFixture('unclosed-braces.astro');
      const result = parseAstro(source);
      
      expect(result.diagnostics).toHaveLength(2);
      expect(result.diagnostics[0].code).toBe('unclosed-expression');
      expect(result.diagnostics[1].code).toBe('unclosed-expression');
      
      const div = result.ast.children[0] as ElementNode;
      const p = div.children.find(child => 
        child.type === 'Element' && (child as ElementNode).tag === 'p'
      ) as ElementNode;
      
      // Should still parse the expression, even if unclosed
      const expr = p.children.find(child => child.type === 'Expression') as ExpressionNode;
      expect(expr).toBeDefined();
      expect(expr.code).toBe('message');
      expect(expr.incomplete).toBe(true);
    });

    test('should handle unbalanced tags with implicit closing', () => {
      const source = loadFixture('unbalanced-tags.astro');
      const result = parseAstro(source);
      
      // Should have diagnostics for unclosed tags
      const unclosedDiagnostics = result.diagnostics.filter(d => d.code === 'unclosed-tag');
      expect(unclosedDiagnostics.length).toBeGreaterThan(0);
      
      const ul = result.ast.children.find(child => 
        child.type === 'Element' && (child as ElementNode).tag === 'ul'
      ) as ElementNode;
      
      // Should have parsed all three li elements
      const liElements = ul.children.filter(child => 
        child.type === 'Element' && (child as ElementNode).tag === 'li'
      );
      expect(liElements).toHaveLength(3);
    });
  });

  describe('duplicate directives', () => {
    test('should warn about duplicate client directives', () => {
      const source = loadFixture('nested-directives.astro');
      const result = parseAstro(source);
      
      const duplicateWarnings = result.diagnostics.filter(d => 
        d.code === 'duplicate-directive' && d.severity === 'warning'
      );
      expect(duplicateWarnings).toHaveLength(2);
      
      // First component with duplicate directives
      const island = result.ast.children.find(child => 
        child.type === 'Component' && (child as ComponentNode).tag === 'Island'
      ) as ComponentNode;
      expect(island.attrs).toHaveLength(2);
      expect(island.attrs[0].name).toBe('client:load');
      expect(island.attrs[1].name).toBe('client:visible');
    });
  });

  describe('location tracking', () => {
    test('should track accurate positions for all nodes', () => {
      const source = `<div>
  <p>Hello</p>
</div>`;
      const result = parseAstro(source);
      
      const div = result.ast.children[0] as ElementNode;
      expect(div.loc.start.line).toBe(1);
      expect(div.loc.start.column).toBe(1);
      
      const p = div.children.find(child => 
        child.type === 'Element' && (child as ElementNode).tag === 'p'
      ) as ElementNode;
      expect(p.loc.start.line).toBe(2);
      expect(p.loc.start.column).toBe(3);
      
      const text = p.children.find(child => child.type === 'Text') as TextNode;
      expect(text.value).toBe('Hello');
    });
  });

  describe('component detection', () => {
    test('should distinguish between HTML elements and components', () => {
      const source = `<div>
  <Button>Click me</Button>
  <button>HTML button</button>
  <MyComponent />
  <my-element />
</div>`;
      const result = parseAstro(source);
      
      const div = result.ast.children[0] as ElementNode;
      const children = div.children.filter(child => 
        child.type === 'Element' || child.type === 'Component'
      );
      
      expect(children[0].type).toBe('Component');
      expect((children[0] as ComponentNode).tag).toBe('Button');
      
      expect(children[1].type).toBe('Element');
      expect((children[1] as ElementNode).tag).toBe('button');
      
      expect(children[2].type).toBe('Component');
      expect((children[2] as ComponentNode).tag).toBe('MyComponent');
      
      expect(children[3].type).toBe('Element');
      expect((children[3] as ElementNode).tag).toBe('my-element');
    });
  });
});