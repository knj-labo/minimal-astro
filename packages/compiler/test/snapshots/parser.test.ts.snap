// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`パーサー > ASTのスナップショットを生成する 1`] = `
{
  "children": [
    {
      "type": "Frontmatter",
      "value": "const items = ['apple', 'banana', 'orange']",
    },
    {
      "attributes": [],
      "children": [
        {
          "attributes": [],
          "children": [
            {
              "attributes": [],
              "children": [
                {
                  "type": "Text",
                  "value": "Fruit List",
                },
              ],
              "name": "title",
              "selfClosing": false,
              "type": "Element",
            },
          ],
          "name": "head",
          "selfClosing": false,
          "type": "Element",
        },
        {
          "attributes": [],
          "children": [
            {
              "attributes": [],
              "children": [
                {
                  "type": "Text",
                  "value": "My Fruits",
                },
              ],
              "name": "h1",
              "selfClosing": false,
              "type": "Element",
            },
            {
              "attributes": [],
              "children": [
                {
                  "type": "Expression",
                  "value": "items.map(item => <li>{item}</li>)",
                },
              ],
              "name": "ul",
              "selfClosing": false,
              "type": "Element",
            },
          ],
          "name": "body",
          "selfClosing": false,
          "type": "Element",
        },
      ],
      "name": "html",
      "selfClosing": false,
      "type": "Element",
    },
  ],
  "type": "Program",
}
`;
